### PR TITLE
chore: Update Templete to Latest Practices

### PR DESCRIPTION
**Pull request by mucahit**
### Description
- All of the package versions are updated
- Added missing `eslint` overrides
- Switched to `react-router-dom`
- Experience improvements in Sentry integration
- Added `hipo-exceptions-js` library
- Pulled latest version of `asyncReduxAutomator` from Pascal Web App

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -29,88 +29,91 @@
...
-    "@types/reach__router": "1.2.4",
-    "@types/react": "16.9.2",
-    "@types/react-dom": "16.9.0",
-    "@types/react-modal": "3.8.2",
-    "@types/react-redux": "7.1.2",
+    "@types/react": "16.9.34",
+    "@types/react-dom": "16.9.6",
+    "@types/react-modal": "3.10.5",
+    "@types/react-redux": "7.1.7",
+    "@types/react-router-dom": "^5.1.4",
```
Comment (edizcelik) [inline package.json]: Should we install the exact version ?

File: src/components/counter/Counter.tsx [hunk 2]
```diff
@@ -2,13 +2,12 @@ import "./_counter.scss";
...
-import {RouteComponentProps} from "@reach/router";
 
-type TCounterProps = {
+interface CounterOwnProps {
   customClassName?: string;
-} & RouteComponentProps;
+}
 
-function Counter({customClassName}: TCounterProps) {
+function Counter({customClassName}: CounterOwnProps) {
```
Comment (edizcelik) [inline src/components/counter/Counter.tsx]: I used to use `CounterOwnProps` when we were using `connect` HOC for `redux`. But as we no longer type the props coming from Redux state, we can directly use `ComponentNameProps`. For connected class components, we can still follow the "own" convention.

```suggestion
function Counter({customClassName}: CounterProps) {
```

File: src/core/redux/reducerFactory.ts [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+import {AsyncActionTypes, TReduxActionWithPayload} from "./models/action";
+import {ApiErrorShape} from "../network-manager/networkModels";
+
+function reduxAsyncReducerFactory<T>(
+  initialState: MinimalAsyncStoreState<T>,
+  actionTypes: AsyncActionTypes,
+  // @ts-ignore
+  type: "basic"
+) {
+  return (state = initialState, action: TReduxActionWithPayload): typeof initialState => {
```
Comment (edizcelik) [inline src/core/redux/reducerFactory.ts]:
```suggestion
  return function asyncReducer (state = initialState, action: TReduxActionWithPayload): typeof initialState {
       let newState = state;
       
       switch (action.type) {
      case actionTypes.REQUEST_TRIGGER: {
        newState = {
          ...state,
          isRequestPending: true,
          isRequestFetched: false
        };
      }
```

File: src/core/route/component/not-found/NotFoundPage.tsx [hunk 4]
```diff
@@ -1,8 +1,7 @@
 import React, {FunctionComponent, useState} from "react";
-import {RouteComponentProps} from "@reach/router";
-import ReportModal from "./modal/components/report/ReportModal";
+import ReportModal from "../../../../components/modal/components/report/ReportModal";
 
-const NotFound: FunctionComponent<RouteComponentProps> = () => {
+const NotFoundPage: FunctionComponent = () => {
```
Comment (edizcelik) [inline src/core/route/component/not-found/NotFoundPage.tsx]: Let's make sure all components in the template follow our practices

```suggestion
function NotFoundPage() {
```

File: src/core/route/publicRoutes.tsx [hunk 5]
```diff
@@ -1,16 +1,15 @@
...
 
 function getPublicRouteComponents() {
   return [
-    <Home
-      key={`route:=${ROUTE_NAMES.ROOT}`}
-      path={ROUTE_NAMES.ROOT}
-      {...ROUTE_CONFIG[ROUTE_NAMES.ROOT].props}
-    />
+    <Route key={`route:=${ROUTE_NAMES.ROOT}`} exact={true} path={ROUTE_NAMES.ROOT}>
+      <Home {...ROUTE_CONFIG[ROUTE_NAMES.ROOT].props} />
```
Comment (edizcelik) [inline src/core/route/publicRoutes.tsx]: I think we should remove `ROUTE_CONFIG` altogether. It is no longer needed and I believe it just adds an unnecessary layer of abstraction.

File: src/core/route/util/routeUtils.ts [hunk 6]
```diff
@@ -0,0 +1,11 @@
+function generateRedirectStateFromLocation(
+  location: any,
```
Comment (edizcelik) [inline src/core/route/util/routeUtils.ts]: We can use `LocationDescriptorObject` type from `history`

```suggestion
  location: LocationDescriptorObject,
```

File: src/core/route/util/routeUtils.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+  shouldOverwriteState?: boolean
```
Comment (edizcelik) [inline src/core/route/util/routeUtils.ts]: Let's remove this argument for now. It isn't used anyway.

File: src/home/Home.tsx [hunk 8]
```diff
@@ -4,15 +4,14 @@ import "./_home.scss";
...
 import {authenticationProfileSelector} from "../authentication/redux/util/authenticationReduxUtils";
 
 export interface HomeOwnProps {
   name: string;
 }
 
-type THomeProps = HomeOwnProps & RouteComponentProps;
+type THomeProps = HomeOwnProps;
 
 function Home(props: THomeProps) {
```
Comment (edizcelik) [inline src/home/Home.tsx]:
```suggestion
function Home(props: HomeProps) {
```

File: src/home/Home.tsx [hunk 9]
```diff
@@ -4,15 +4,14 @@ import "./_home.scss";
 
 import React from "react";
 import {useSelector} from "react-redux";
-import {RouteComponentProps} from "@reach/router";
 
 import {authenticationProfileSelector} from "../authentication/redux/util/authenticationReduxUtils";
 
 export interface HomeOwnProps {
```
Comment (edizcelik) [inline src/home/Home.tsx]: We no longer need to export this, as we wont be using routeConfig
```suggestion
interface HomeProps {
```

File: src/home/Home.tsx [hunk 10]
```diff
@@ -4,15 +4,14 @@ import "./_home.scss";
...
   name: string;
 }
 
-type THomeProps = HomeOwnProps & RouteComponentProps;
+type THomeProps = HomeOwnProps;
```
Comment (edizcelik) [inline src/home/Home.tsx]: Let's remove this line

File: src/core/redux/reducerFactory.ts [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+          data: action.payload,
+          errorInfo: null
+        };
+      }
+
+      case actionTypes.REQUEST_ERROR: {
+        let finalErrorInfo: null | ApiErrorShape = null;
+
+        if (action.payload instanceof Error) {
+          finalErrorInfo = {
```
Comment (edizcelik) [inline src/core/redux/reducerFactory.ts]: Let's use `createErrorWithApiErrorShapeFromString` utility

**Comment by edizcelik** [comment]
Can you also add `describe` to the basic tests ?
https://github.com/Hipo/hipo-frontend-project-starter/issues/4

**Comment by edizcelik** [comment]
Let's add CircleCI config to the template:

```
version: 2
jobs:
  deploy_staging:
    docker:
      - image: circleci/node:12.16.1
    steps:
      - checkout
      - run: npm ci --no-progress --prefer-offline
      - run: npm run build:release -- --env.target=staging --env.awsAccessKeyId=${AWS_STAGING_ACCESS_KEY} --env.awsSecretAccessKey=${AWS_STAGING_SECRET_ACCESS_KEY}
      # - run:
      #     name: Run QA tests
      #     command: curl -i -X POST 'http://testrunner.hipolabs.com/job/manufacturit-web/build?token=manufacturit' --user ${TEST_RUNNER_USERNAME}:${TEST_RUNNER_PASSWORD}
  deploy_production:
    docker:
      - image: circleci/node:12.16.1
    steps:
      - checkout
      - run: npm ci --no-progress --prefer-offline
      - run: npm run build:release -- --env.target=production --env.awsAccessKeyId=${AWS_PRODUCTION_ACCESS_KEY} --env.awsSecretAccessKey=${AWS_PRODUCTION_SECRET_ACCESS_KEY}
workflows:
  version: 2
  build:
    jobs:
      - deploy_staging:
          filters:
            branches:
              only: staging
      - deploy_production:
          filters:
            branches:
              only: production
```

We can comment out `deploy_production` job.
https://github.com/Hipo/hipo-frontend-project-starter/issues/3

**Comment by mucahit** [comment]
Thanks for the review @edizcelik

**Review by edizcelik — APPROVED** [review]

**Review by burakguneli — APPROVED** [review]

**Pull request merged**